File: python/354_RussianDollEnvelopes.py
from unittest import TestCase
# https://leetcode.com/problems/russian-doll-envelopes/


class RussianDollEnvelopes(TestCase):
    def maxEnvelopes(self, envelopes):
        """
        :type envelopes: List[List[int]]
        :rtype: int
        """
        n = len(envelopes)
        ret = 0
        mins = [-1, *[float("inf")]*(n+1)]
        for w, h in sorted(envelopes, key=lambda x: (x[0], -x[1])):
            l, r = 0, n
            while l <= r:
                m = int((l+r)/2)
                if mins[m] < h:
                    l = m + 1
                else:
                    r = m - 1
            mins[l] = h
            ret = max(ret, l)
        return ret

    def test1(self):
        self.assertEqual(3, self.maxEnvelopes([[5,4],[6,4],[6,7],[2,3]]))

File: kotlin/src/main/kotlin/354.RussianDollEnvelopes.kt
import org.junit.Assert.assertEquals
import org.junit.Test
//https://leetcode.com/problems/russian-doll-envelopes/


class RussianDollEnvelopes{
  fun maxEnvelopes(envelopes: Array<IntArray>): Int {
    var ret = 0
    val sorted = envelopes.sortedWith(compareBy({it[0]}, {-it[1]})).map{it[1]}
    val n = envelopes.size
    val mins = IntArray(n+1){ Int.MAX_VALUE }
    mins[0] = Int.MIN_VALUE

    for(h in sorted){
      var (l, r) = 0 to n
      while(l <= r){
        val m = (l+r)/2
        if(mins[m] < h)
          l = m+1
        else
          r = m-1
      }
      mins[l] = h
      ret = maxOf(ret, l)
    }
    return ret
  }

  @Test
  fun test1(){
    assertEquals(3, maxEnvelopes(arrayOf(
        intArrayOf(5,4),
        intArrayOf(6,4),
        intArrayOf(6,7),
        intArrayOf(2,3))))
  }
}